src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -375,185 +366,200 @@
-
-        ;; initial queue population
-        (future
-          (println "Populating the queue in the background")
-          (dotimes [n numhosts]
-            (let [h (-> (make-host n)
-                        (assoc :lastrun (rand-lastrun run-interval)))]
-              (>!! mq h)))
-
-          (println "... finished filling the queue, continuing with the simulation."))
-
-        (let [run-interval-minutes (time/in-minutes run-interval)
-              hosts-per-second (/ numhosts (* run-interval-minutes 60))
-              rate-limited-mq-out (chan)]
-          (relay-messages-at-rate mq rate-limited-mq-out hosts-per-second)
-          (pipeline simulation-threads
-                    mq
-                    (map #(update-host % rand-perc (time/now) command-send-ch))
-                    rate-limited-mq-out)
-          (go
-            (<! close-to-stop-ch)
-            (mapv close! [rate-limited-mq-out mq rate-monitor-ch command-send-ch])))
-        close-to-stop-ch))))
-
-(defn chan? [x]
-  (satisfies? clojure.core.async.impl.protocols/Channel x))
-
+                                   (if facts 1 0))
+        temp-dir (get-path (or (System/getenv "TMPDIR")
+                               (System/getProperty "java.io.tmpdir")))
+
+        ;; channels
+        mq-ch (chan (message-buffer temp-dir numhosts))
+        _ (.addShutdownHook (Runtime/getRuntime) (Thread. #(async/close! mq-ch)))
+
+        command-send-ch (chan)
+        rate-monitor-ch (chan)
+
+        ;; processes
+        rate-monitor-finished-ch (start-rate-monitor rate-monitor-ch run-interval commands-per-puppet-run)
+        command-sender-finished-ch (start-command-sender base-url
+                                                         (if nummsgs
+                                                           (async/take (* numhosts nummsgs) command-send-ch)
+                                                           command-send-ch)
+                                                         rate-monitor-ch
+                                                         threads)
+        populate-finished-ch (start-populate-queue mq-ch numhosts run-interval pdb-host
+                                                   catalogs reports facts)
+        _ (start-simulation-loop numhosts run-interval nummsgs rand-perc
+                                 simulation-threads command-send-ch mq-ch)
+        join-fn (fn join-benchmark
+                  ([] (join-benchmark nil))
+                  ([timeout-ms]
+                   (let [t-ch (if timeout-ms (async/timeout timeout-ms) (chan))]
+                     (async/alt!! t-ch false populate-finished-ch true)
+                     (async/alt!! t-ch false command-sender-finished-ch true)
+                     (async/alt!! t-ch false rate-monitor-ch true))))
+        stop-fn (fn stop-benchmark []
+                  (async/close! mq-ch)
+                  (join-fn 5000))
```

Comment:
join (and stop) actually do that :) they return true on successful join, or false on timeout.